Add caveat details in auth manager doc (#53397)
mistercrunch/Airflow: airflow-core/docs/core-concepts/auth-manager/index.rst

  * ``POST``: Can the user create a resource?
  * ``PUT``: Can the user modify the resource?
  * ``DELETE``: Can the user delete the resource?

>   * ``MENU``: Can the user see the resource in the menu? Note: this method is only applicable to a subset of auth manager Authorization methods.

Reviewer: This has been removed recently actually. Can you remove `MENU` from the list?

Author: Sounds good. That reminds me -- I had a question on this point, should I make note of the newer [`ExtendedResourceMethod`](https://github.com/apache/airflow/blob/main/airflow-core/src/airflow/api_fastapi/auth/managers/base_auth_manager.py#L67-L68) definition that was recently added? Or is that still under development?

Reviewer: It is no longer under development so feel free to mention it

* ``is_authorized_custom_view``: Return whether the user is authorized to access a specific view not defined in Airflow. This view can be provided by the auth manager itself or a plugin defined by the user.
* ``filter_authorized_menu_items``: Given the list of menu items in the UI, return the list of menu items the user has access to.

It should be noted that the parameter ``methods`` listed above may only have relevance for a specific subset of the auth manager's authorization methods.

> For example, the ``configuration`` resource is by definition read-only, so only the ``GET`` parameter method is relevant in the context of ``is_authorized_configuration``.

Reviewer: ```suggestion
For example, the ``configuration`` resource is by definition read-only, so only the ``GET`` parameter is relevant in the context of ``is_authorized_configuration``.
```

Author: added 👍

* ``is_authorized_custom_view``: Return whether the user is authorized to access a specific view not defined in Airflow. This view can be provided by the auth manager itself or a plugin defined by the user.
* ``filter_authorized_menu_items``: Given the list of menu items in the UI, return the list of menu items the user has access to.

It should be noted that the API ``methods`` listed above may only have relevance for a specific subset of the auth manager's authorization methods.

> For example, the ``configuration`` resource is by definition read-only, so only the ``GET`` API method is relevant in the context of ``is_authorized_configuration``.

Reviewer: ```suggestion
It should be noted that the parameter ``methods`` listed above may only have relevance for a specific subset of the auth manager's authorization methods.
For example, the ``configuration`` resource is by definition read-only, so only the ``GET`` parameter method is relevant in the context of ``is_authorized_configuration``.
```

JWT token management by auth managers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The auth manager is responsible for creating the JWT token needed to interact with Airflow public API.
To achieve this, the auth manager **must** provide an endpoint to create this JWT token. This endpoint is usually
available at ``POST /auth/token``.
Please double check the auth manager documentation to find the accurate token generation endpoint.

The auth manager is also responsible for passing the JWT token to Airflow UI. The protocol to exchange the JWT
token between the auth manager and Airflow UI is using cookies. The auth manager needs to save the JWT token in a
cookie named ``_token`` before redirecting to the Airflow UI. The Airflow UI will then read the cookie, save it, and delete it.

.. code-block:: python

    from airflow.api_fastapi.auth.managers.base_auth_manager import COOKIE_NAME_JWT_TOKEN

    response = RedirectResponse(url="/")

>     secure = request.base_url.scheme == "https" or bool(conf.get("api", "ssl_cert", fallback=""))

Author: Let me know if this is no longer the case, but judging by the source code in the various auth managers, I think this is the proper way to determine the `secure` value for the JWT cookie. I was thinking of putting up a separate PR to abstract this repeat logic into a common util function.

Reviewer: That looks good to me

* ``batch_is_authorized_dag``: Batch version of ``is_authorized_dag``. If not overridden, it will call ``is_authorized_dag`` for every single item.
* ``get_authorized_dag_ids``: Return the list of DAG IDs the user has access to.  If not overridden, it will call ``is_authorized_dag`` for every single DAG available in the environment.

>   * Note: To filter the results of ``get_authorized_dag_ids``, it is recommended that you define the filtering logic in your ``filter_authorized_dag_ids`` method. For example, this may be usefule if you rely on per-DAG access controls derived from one or more fields on a given DAG (e.g. DAG tags).

Reviewer: Nice catch from the spellcheck

```suggestion
  * Note: To filter the results of ``get_authorized_dag_ids``, it is recommended that you define the filtering logic in your ``filter_authorized_dag_ids`` method. For example, this may be useful if you rely on per-DAG access controls derived from one or more fields on a given DAG (e.g. DAG tags).
```

Author: oof, my bad for missing this -- I completely forgot that I had left my pre-commit setup disabled while working on this 🤦‍♂️. Fixed this (and re-enabled pre-commit). Sorry for the added noise here!

* ``batch_is_authorized_dag``: Batch version of ``is_authorized_dag``. If not overridden, it will call ``is_authorized_dag`` for every single item.
* ``get_authorized_dag_ids``: Return the list of DAG IDs the user has access to.  If not overridden, it will call ``is_authorized_dag`` for every single DAG available in the environment.

>   * Note: The ``get_authorized_dag_ids`` may be of particular interest if you rely on per-DAG access controls derived from one or more fields on a given DAG (e.g. DAG tags).

Reviewer: Actually, thanks for catching this but `get_authorized_dag_ids` is not supposed to be overridden by users. On the other side, `filter_authorized_dag_ids` is. `get_authorized_dag_ids` retrieve all the DAGs from the DB and then call `filter_authorized_dag_ids` to filter DAGs based on user permissions.

Author: Oh that's interesting, thank you for bringing that to my attention. Are you saying that the `get_authorized_dag_ids` should not be considered part of the public Auth Manager interface? If so, I'd be interested in helping make that an optionally overridable component of the interface. The reason being that it could allow for optimizations at the query level, while also giving more expressive options for DAG ownership inference via DAG attributes (again, like DAG tags).

Reviewer: No, `get_authorized_dag_ids` should be part of the auth manager interface. In that sense, users are free to override it but they are more expected to not override it. On the other side `filter_authorized_dag_ids`, is more expected to be overridden. But you are right, if they want to, users are free to override `get_authorized_dag_ids` to make optimizations on the query level, at their own risks.

Author: Got it, thanks for the clarification. I went ahead and slightly reworded this section. Please let me know what you think. I left the `get_authorized_dag_ids` mention in this section since the section is still technically related to optimizations, but added a clarification about overriding `get_authorized_dag_ids` vs overriding just `filter_authorized_dag_ids`.

  their class paths need to be returned as part of this method. By doing so, they will be automatically added to the
  config ``[database] external_db_managers``.

Additional Caveats
^^^^^^^^^^^^^^^^^^

* Your auth manager should not reference anything from the ``airflow.security.permissions`` module, as that module is in the process of being deprecated.
  * Instead, your code should use the definitions in ``airflow.api_fastapi.auth.managers.models.resource_details``.
* The ``access_control`` attribute of a DAG instance is only compatible with the FAB auth manager. Custom auth manager implementations should leverage ``get_authorized_dag_ids`` for DAG instance attribute-based access controls in more customizable ways (e.g. authorization based on DAG tags, DAG bundles, etc.).

> DAG and DAG Sub-Component Authorization

Reviewer: Nice section! It gives some nice details which can helped to understand better the permission model 👍
